src/main/java/ir/sahab/kafkaconsumer/OffsetTracker.java

```diff
@@ -119,51 +120,70 @@ public OptionalLong ack(int partition, long offset) {
                     (Gauge<Integer>) () -> openPagesSize);
             metricRegistry.register(LabeledMetric.name(COMPLETED_PAGES).label(PARTITION, partitionLabel).toString(),
                     (Gauge<Integer>) () -> completedPagesSize);
+
+            openNewPageForOffset(initialOffset);
+            lastCompletedPageIndex = offsetToPage(initialOffset) - 1;
+            lastTrackedOffset = initialOffset - 1;
         }
 
         boolean track(long offset) {
             long pageIndex = offsetToPage(offset);
-            int pageOffset = (int) (offset % pageSize);
-            PageTracker pageTracker = pageTrackers.get(pageIndex);
-            if (pageTracker == null) {
-                if (pageTrackers.size() >= maxOpenPagesPerPartition) {
+
+            // Normal case: offsets coming in order.
+            if (offset - lastTrackedOffset == 1) {
+                // If offset belongs to new page and we can't create a new page, return false.
+                if (pageIndex > lastOpenedPageIndex && !openNewPageForOffset(offset)) {
                     return false;
                 }
-                pageTracker = new PageTracker(pageSize, pageOffset);
-                pageTrackers.put(pageIndex, pageTracker);
-                openPagesSize = pageTrackers.size();
-                // Since all the offsets will be tracked in order, we don't expect any more tracks for previous page.
-                markPageNoMoreTracks(pageIndex - 1);
+                lastTrackedOffset = offset;
+                return true;
+            } else if (offset <= lastTrackedOffset) {
+                return true; // Old offsets buffered before rebalance.
+            }
+
+            // Special cases: we have seen a offset greater than expectation and there is a gap between offsets.
+            final PageTracker lastPage = pageTrackers.get(lastOpenedPageIndex);
+            if (pageIndex == lastOpenedPageIndex) { // The gap is in the same page. Fill it and assume it's acked before.
+                    lastPage.bulkAck(offsetToMargin(lastTrackedOffset), offsetToMargin(offset));
+            } else { // The gap includes some pages in between and we must open a new page for the new offset.
+                final long lastPageIndexBeforeGap = lastOpenedPageIndex;
+                final long lastOffsetBeforeGap = lastTrackedOffset;
+                if (!openNewPageForOffset(offset)) return false;
+                // Reduce the size of last page and check if it's completed.
+                boolean completed = lastPage.setSize(offsetToMargin(lastOffsetBeforeGap));
+                if (completed) {
+                    pageTrackers.remove(lastPageIndexBeforeGap);
+                    completedPages.add(lastPageIndexBeforeGap);
+                }
+                // Assume missing pages are completed before.
+                LongStream.range(lastPageIndexBeforeGap + 1, pageIndex).forEach(completedPages::add);
```

Comment:
Here you add pages in the gap as completed pages, but you should be careful, in `ack()` method, that we should not return an index that belongs to these pages.